addons/mail/static/src/models/thread_viewer/thread_viewer.js

```diff
@@ -131,75 +126,30 @@ function factory(dependencies) {
 
         /**
          * @private
-         * @returns {integer|undefined}
-         */
-        _computeThreadCacheInitialScrollPosition() {
-            if (!this.threadCache) {
-                return undefined;
-            }
-            return this.threadCacheInitialScrollPositions[this.threadCache.localId];
-        }
-
-        /**
-         * @private
-         */
-        _prepareLoading() {
-            this._isPreparingLoading = true;
-            this._loaderTimeout = setTimeout(() => {
-                this.isShowingLoading = true;
-                this._isPreparingLoading = false;
-            }, 400);
-        }
-
-        /**
-         * @private
-         */
-        _stopLoading() {
-            clearTimeout(this._loaderTimeout);
-            this._loaderTimeout = null;
-            this.isShowingLoading = false;
-            this._isPreparingLoading = false;
-        }
-
-        /**
-         * @override
          */
-        _updateAfter(previous) {
-            if (this.thread && this.thread !== previous.thread) {
-                this._stopLoading();
-                if (!this.threadCache.isLoaded && !this.threadCache.isLoading) {
-                    this.threadCache.loadMessages();
+        _onThreadCacheIsLoadingChanged() {
+            if (this.threadCache && this.threadCache.isLoading) {
+                if (!this.isLoading && !this.isPreparingLoading) {
+                    this.update({ isPreparingLoading: true });
+                    this.async(() =>
+                        new Promise(resolve => {
+                            this._loaderTimeout = setTimeout(resolve, 400);
+                        }
+                    )).then(() => {
+                        const isLoading = this.threadCache
+                            ? this.threadCache.isLoading
+                            : false;
+                        this.update({ isLoading, isPreparingLoading: false });
+                    });
                 }
+                return;
             }
-            if (this.threadCache !== previous.threadCache) {
-                this._stopLoading();
-                this.addComponentHint('change-of-thread-cache');
-            }
-
-            if (
-                this.thread && this.threadCache && this.threadCache.isLoading &&
-                !this.isShowingLoading && !this._isPreparingLoading
-            ) {
-                this._prepareLoading();
-            }
-        }
-
-        /**
-         * @override
-         */
-        _updateBefore() {
-            return {
-                thread: this.thread,
-                threadCache: this.threadCache,
-            };
+            clearTimeout(this._loaderTimeout);
```

Comment:
`this.env.browser.setTimeout` and `this.env.browser.clearTimeout()`

Reply:
good catch